src/main/java/net/dries007/tfc/common/blocks/wood/VerticalSupportBlock.java

```diff
@@ -0,0 +1,163 @@
+    public ForgeBlockProperties getForgeProperties()
+    {
+        return properties;
+    }
+
+    @Override
+    public void setPlacedBy(World worldIn, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack stack)
+    {
+        if (worldIn.isClientSide() || placer == null) return;
+        if (stack.getCount() > 2 && !placer.isShiftKeyDown()) // need two because the item block hasn't shrunk the stack yet
+        {
+            BlockPos above = pos.above();
+            BlockPos above2 = above.above();
+            if (worldIn.isEmptyBlock(above) && worldIn.isEmptyBlock(above2))
+            {
+                if (worldIn.noCollision(placer, new AxisAlignedBB(above)))
+                {
+                    worldIn.setBlock(above, defaultBlockState(), 2);
+                    if (worldIn.noCollision(placer, new AxisAlignedBB(above2)))
+                    {
+                        worldIn.setBlock(above2, defaultBlockState(), 2);
+                        stack.shrink(2);
+                    }
+                    else
+                    {
+                        stack.shrink(1);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    @Nullable
+    public BlockState getStateForPlacement(BlockItemUseContext context)
+    {
+        BlockState state = defaultBlockState();
+        BlockPos.Mutable mutablePos = new BlockPos.Mutable();
+        for (Direction d : Direction.Plane.HORIZONTAL)
+        {
+            mutablePos.setWithOffset(context.getClickedPos(), d);
+            state = state.setValue(PROPERTY_BY_DIRECTION.get(d), context.getLevel().getBlockState(mutablePos).is(TFCTags.Blocks.SUPPORT_BEAM));
+        }
+        return state;
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public BlockState updateShape(BlockState stateIn, Direction facing, BlockState facingState, IWorld world, BlockPos currentPos, BlockPos facingPos)
+    {
+        if (facing.getAxis().isHorizontal())
+        {
+            stateIn = stateIn.setValue(PROPERTY_BY_DIRECTION.get(facing), facingState.is(TFCTags.Blocks.SUPPORT_BEAM));
+        }
+        else if (facing == Direction.DOWN)
+        {
+            if (facingState.is(TFCTags.Blocks.SUPPORT_BEAM) || facingState.isFaceSturdy(world, facingPos, Direction.UP, BlockVoxelShape.CENTER))
+            {
+                return stateIn;
+            }
+            return Blocks.AIR.defaultBlockState();
+        }
+        return stateIn;
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public VoxelShape getShape(BlockState state, IBlockReader worldIn, BlockPos pos, ISelectionContext context)
+    {
+        VoxelShape shape = getMiddleShape();
+        for (Direction d : Direction.Plane.HORIZONTAL)
+        {
+            if (state.getValue(PROPERTY_BY_DIRECTION.get(d)))
+            {
+                shape = VoxelShapes.or(shape, shapeByDirection(d));
+            }
+        }
+        return shape;
```

Comment:
Since `VoxelShape` unions can be relatively expensive (and `getShape` called a lot), it's good practice to cache these, either from property or the state. For example, see `WallBlock#makeShapes` in vanilla - it builds a `Map<BlockState, VoxelShape>` and then that is used in this method to query.

Additionally, because of this, `getMiddleShape` need not be an overridable method, it can become a constructor parameter since this is it's only practical usage.